app/bundles/PageBundle/Tests/EventListener/SegmentTrackingSubscriberFunctionalTest.php

```diff
@@ -0,0 +1,410 @@
+ */
+final class SegmentTrackingSubscriberFunctionalTest extends MauticMysqlTestCase
+{
+    private const TEST_URL = 'https://example.com/page';
+
+    private const SEGMENT_IDS_PARAM = 'segment_ids=';
+
+    private EventDispatcherInterface $dispatcher;
+
+    protected function setUp(): void
+    {
+        $this->configParams['append_segment_id_tracking_url'] = 'testFeatureFlagIsDisabled' !== $this->name();
+        parent::setUp();
+        $this->dispatcher = $this->getContainer()->get('event_dispatcher');
+    }
+
+    public function testFeatureFlagIsDisabled(): void
+    {
+        $lead      = $this->createContactWithSegments(2);
+        $resultUrl = $this->dispatchUrlTokenReplaceEvent(self::TEST_URL, $lead->getId());
+
+        Assert::assertSame(self::TEST_URL, $resultUrl, 'When feature flag is disabled, URL should not be modified');
+        Assert::assertStringNotContainsString('segment_ids', $resultUrl, 'URL should not contain segment_ids parameter when feature flag is disabled');
+    }
+
+    public function testFeatureFlagIsEnabled(): void
+    {
+        $lead      = $this->createContactWithSegments(2);
+        $resultUrl = $this->dispatchUrlTokenReplaceEvent(self::TEST_URL, $lead->getId());
+
+        Assert::assertStringContainsString(self::SEGMENT_IDS_PARAM, $resultUrl, 'When feature flag is enabled, segment IDs should be appended');
+        Assert::assertStringStartsWith(self::TEST_URL.'?'.self::SEGMENT_IDS_PARAM, $resultUrl, 'Segment IDs should be appended as query parameter');
+    }
+
+    #[\PHPUnit\Framework\Attributes\DataProvider('provideContactInputTypes')]
+    public function testSegmentIdsAppendedWithDifferentContactTypes(string $testCase, callable $getContact): void
+    {
+        $lead        = $this->createContactWithSegments(3);
+        $segments    = $this->getContactSegments($lead);
+        $expectedIds = array_keys($segments);
+        sort($expectedIds);
+
+        $contact   = $getContact($lead);
+        $resultUrl = $this->dispatchUrlTokenReplaceEvent(self::TEST_URL, $contact);
+
+        Assert::assertStringContainsString(self::SEGMENT_IDS_PARAM, $resultUrl, $testCase);
+
+        $parsedUrl = parse_url($resultUrl);
+        parse_str($parsedUrl['query'], $queryParams);
+
+        Assert::assertArrayHasKey('segment_ids', $queryParams);
+
+        $actualIds = explode(',', $queryParams['segment_ids']);
+        sort($actualIds);
+
+        Assert::assertSame($expectedIds, array_map('intval', $actualIds), $testCase);
+    }
+
+    /**
+     * @return iterable<string, array{string, callable}>
+     */
+    public static function provideContactInputTypes(): iterable
+    {
+        yield 'Contact as integer ID' => [
+            'Integer contact ID should work',
+            fn (Lead $lead) => $lead->getId(),
+        ];
+
+        yield 'Contact as Lead entity' => [
+            'Lead entity should work',
+            fn (Lead $lead): Lead => $lead,
+        ];
+    }
```

Comment:
I think these tests are testing an artificial scenarios that will never happen when calling the actual redirect route. It would be much better to simply call the redirect route with various input that can happen and various states in the database.